pkg/types/dsse/v0.0.1/entry.go

```diff
@@ -0,0 +1,351 @@
+	}
+
+	allPubKeyBytes := make([][]byte, 0)
+	if props.PublicKeyBytes != nil {
+		allPubKeyBytes = append(allPubKeyBytes, props.PublicKeyBytes)
+	}
+
+	allPubKeyBytes = append(allPubKeyBytes, props.PublicKeysBytes...)
+	allPubKeyPaths := make([]*url.URL, 0)
+	if props.PublicKeyPath != nil {
+		allPubKeyPaths = append(allPubKeyPaths, props.PublicKeyPath)
+	}
+
+	for _, path := range allPubKeyPaths {
+		if path.IsAbs() {
+			return nil, errors.New("dsse public keys cannot be fetched over HTTP(S)")
+		}
+
+		publicKeyBytes, err := ioutil.ReadFile(filepath.Clean(path.Path))
+		if err != nil {
+			return nil, fmt.Errorf("error reading public key file: %w", err)
+		}
+
+		allPubKeyBytes = append(allPubKeyBytes, publicKeyBytes)
+	}
+
+	keysBySig, err := verifyEnvelope(allPubKeyBytes, &env)
+	if err != nil {
+		return nil, err
+	}
+
+	decodedPayload, err := base64.StdEncoding.DecodeString(env.Payload)
+	if err != nil {
+		return nil, fmt.Errorf("could not decode envelope payload: %w", err)
+	}
+
+	paeEncodedPayload := dsse.PAE(env.PayloadType, decodedPayload)
+	h := sha256.Sum256(paeEncodedPayload)
+	re.DsseObj.Payload = strfmt.Base64(env.Payload)
+	re.DsseObj.PayloadType = &env.PayloadType
+	re.DsseObj.PayloadHash = &models.DsseV001SchemaPayloadHash{
+		Algorithm: models.DsseV001SchemaPayloadHashAlgorithmSha256,
+		Value:     hex.EncodeToString(h[:]),
+	}
+
+	for _, sig := range env.Signatures {
+		key, ok := keysBySig[sig.Sig]
+		if !ok {
+			return nil, errors.New("all signatures must have a key that verifies it")
+		}
+
+		canonKey, err := key.CanonicalValue()
+		if err != nil {
+			return nil, fmt.Errorf("could not canonicize key: %w", err)
+		}
+
+		keyBytes := strfmt.Base64(canonKey)
+		re.DsseObj.Signatures = append(re.DsseObj.Signatures, &models.DsseV001SchemaSignaturesItems0{
+			Keyid:     sig.KeyID,
+			Sig:       sig.Sig,
+			PublicKey: keyBytes,
+		})
+	}
+
+	returnVal.APIVersion = swag.String(re.APIVersion())
+	returnVal.Spec = re.DsseObj
+	return &returnVal, nil
+}
+
+// verifyEnvelope takes in an array of possible key bytes and attempts to parse them as x509 public keys.
+// it then uses these to verify the envelope and makes sure that every signature on the envelope is verified.
+// it returns a map of verifiers indexed by the signature the verifier corresponds to.
+func verifyEnvelope(allPubKeyBytes [][]byte, env *dsse.Envelope) (map[string]*x509.PublicKey, error) {
+	// generate a fake id for these keys so we can get back to the key bytes and match them to their corresponding signature
+	verifierBySig := make(map[string]*x509.PublicKey)
+	allSigs := make(map[string]struct{})
+	for _, sig := range env.Signatures {
+		allSigs[sig.Sig] = struct{}{}
+	}
+
+	for _, pubKeyBytes := range allPubKeyBytes {
+		key, err := x509.NewPublicKey(bytes.NewReader(pubKeyBytes))
+		if err != nil {
+			return nil, fmt.Errorf("could not parse public key as x509: %w", err)
+		}
+
+		vfr, err := signature.LoadVerifier(key.CryptoPubKey(), crypto.SHA256)
```

Comment:
is the hash algorithm always going to be SHA256 for this?

Reply:
While not part of the DSSE spec it seems that all current users of DSSE default to using SHA256 during signing.